### ISSUE TITLE
Use a filter callback instead of list of tokens for TokenBias

### ISSUE THREAD

**Issue by pixelspark** [rustformers/llm#308]
In some cases it is useful to bias a model's output to a large set of tokens (i.e. all tokens that are just ASCII). Currently you'd have to filter all tokens and create a `Vec<(TokenId,f32)>` containing a potentially large number of elements.

My suggested solution would be to replace `TokenBias` with a trait that allows the user to 'filter' tokens:

````rust
pub trait TokenBias {
  fn bias_for_token(&self, token: TokenId, vocabulary: &Vocabulary) -> f32;
}
````

A sample implementation allowing only ASCII:

````rust
struct AsciiTokenBias {}

impl TokenBias for AsciiTokenBias {
  fn bias_for_token(&self, token: TokenId, vocabulary: &Vocabulary) -> f32 {
     let s = String::from_utf8_lossy(&vocabulary.decode(vec![token], false));
     s.contains_only_ascii() // hypothetical
  }
}
````

A default implementation could be provided that accepts a `Vec<(TokenId, f32)>` and simply looks up the associated bias value (using a `HashMap` behind the scenes would be a bit faster I presume).

The sampler need only call the `bias_for_token` function once for each unique TokenId it encounters (and could possibly cache the result).

**Comment by pixelspark** [comment]
Having looked at this some more, one other issue I encounter with the current `TopPTopK` sampler is that the biases provided actually *replace* the predicted logits, whereas for biasing towards a specific output (e.g. ASCII-only) you would want to fully remove certain tokens from the set (and/or possibly adjust logits by adding/subtracing). I am probably better off re-implementing the TopPTopK sampler for this use case. 

The above proposal could still be useful, but then there should be a way to indicate how the bias should be applied (either 'override', 'scale' or 'add'?).
